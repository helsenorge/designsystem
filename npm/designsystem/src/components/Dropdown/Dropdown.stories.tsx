import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Dropdown, { DropdownMode } from './Dropdown';
import Checkbox from '../Checkbox';
import RadioButton from '../RadioButton';
import GridExample from '../GridExample';

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
  argTypes: {
    label: {
      control: 'text',
      defaultValue: 'Ta et valg',
    },
    placeholder: {
      control: 'text',
      defaultValue: 'Hva skjer i kroppen?',
    },
    closeText: {
      control: 'text',
      defaultValue: 'Lukk',
    },
    open: {
      control: 'boolean',
      defaultValue: false,
    },
    mode: {
      control: 'select',
      options: DropdownMode,
      defaultValue: DropdownMode.onwhite,
    },
    transparent: {
      control: 'boolean',
      defaultValue: false,
    },
    fluid: {
      control: 'boolean',
      defaultValue: false,
    },
    noCloseButton: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof Dropdown>;

export const RadioButtonChildren: ComponentStory<typeof Dropdown> = (args: any) => (
  <GridExample>
    <Dropdown {...args} onToggle={action('onToggle')}>
      <RadioButton label="Valg 1" name="radiobutton" />
      <RadioButton label="Valg 2" name="radiobutton" />
      <RadioButton label="Valg 3" name="radiobutton" />
      <RadioButton label="Valg 4" name="radiobutton" />
      <RadioButton label="Valg 5" name="radiobutton" />
      <RadioButton label="Valg 6" name="radiobutton" />
      <RadioButton label="Valg 7" name="radiobutton" />
      <RadioButton label="Valg 8" name="radiobutton" />
      <RadioButton label="Valg 9" name="radiobutton" />
    </Dropdown>
  </GridExample>
);

export const CheckboxChildren: ComponentStory<typeof Dropdown> = (args: any) => (
  <GridExample>
    <Dropdown {...args} onToggle={action('onToggle')}>
      <Checkbox label="Valg 1" name="checkbox" />
      <Checkbox label="Valg 2" name="checkbox" />
      <Checkbox label="Valg 3" name="checkbox" />
      <Checkbox label="Valg 4" name="checkbox" />
      <Checkbox label="Valg 5" name="checkbox" />
      <Checkbox label="Valg 6" name="checkbox" />
      <Checkbox label="Valg 7" name="checkbox" />
      <Checkbox label="Valg 8" name="checkbox" />
      <Checkbox label="Valg 9" name="checkbox" />
    </Dropdown>
  </GridExample>
);

export const CustomContent: ComponentStory<typeof Dropdown> = (args: any) => (
  <GridExample>
    <Dropdown {...args} onToggle={action('onToggle')}>
      <div style={{ padding: '1rem' }}>
        {
          'Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen?'
        }
      </div>
    </Dropdown>
  </GridExample>
);
