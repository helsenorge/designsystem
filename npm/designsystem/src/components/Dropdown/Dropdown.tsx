import React, { useRef, useState } from 'react';

import classNames from 'classnames';

import {
  AnalyticsId,
  IconSize,
  KeyboardEventKey,
  ZIndex,
  theme,
  useHover,
  useKeyboardEvent,
  useOutsideEvent,
  useToggle,
  useUuid,
} from '../..';
import { mergeRefs } from '../../utils/refs';
import Button from '../Button';
import Icon from '../Icon';
import PlusSmall from '../Icons/PlusSmall';

import styles from './styles.module.scss';

export enum DropdownMode {
  onwhite = 'onwhite',
  ongrey = 'ongrey',
  onblueberry = 'onblueberry',
  oncherry = 'oncherry',
}

export interface DropdownProps {
  /** Label for dropdown. Visible for screen readers  */
  label: string;
  /** Text on the trigger button that opens the dropdown */
  placeholder: string;
  /** Sets the dropdown content */
  children: React.ReactNode;
  /** Close button text */
  closeText?: string;
  /** No close button */
  noCloseButton?: boolean;
  /** Called when dropdown is open/closed */
  onToggle?: (isOpen: boolean) => void;
  /** Whether the dropdown is open or not */
  open?: boolean;
  /** Changes the visuals of the dropdown */
  mode?: keyof typeof DropdownMode;
  /** Makes the background of the trigger transparent */
  transparent?: boolean;
  /** Makes the width of the full component adjust to its parent */
  fluid?: boolean;
  /** Makes the dropdown disabled */
  disabled?: boolean;
  /** Sets the data-testid attribute on the dropdown button */
  testId?: string;
  /** Overrides the default z-index of the DropDownContent */
  zIndex?: number;
}

const Dropdown: React.FC<DropdownProps> = props => {
  const {
    label,
    placeholder,
    closeText = 'Lukk',
    noCloseButton = false,
    onToggle,
    open = false,
    children,
    mode = DropdownMode.onwhite,
    transparent = false,
    fluid = false,
    testId,
    disabled,
    zIndex = ZIndex.PopOver,
  } = props;
  const dropdownRef = useRef<HTMLDivElement>(null);
  const optionsRef = useRef<HTMLUListElement>(null);
  const { hoverRef: buttonRef, isHovered } = useHover<HTMLButtonElement>();
  const { value: isOpen, toggleValue: toggleIsOpen } = useToggle(!disabled && open, onToggle);
  const inputRefList = useRef(React.Children.map(children, () => React.createRef<HTMLElement>()));
  const [currentIndex, setCurrentIndex] = useState<number>();
  const labelId = useUuid();
  const toggleLabelId = useUuid();
  const optionIdPrefix = useUuid();

  const handleOpen = (): void => {
    toggleIsOpen();
    optionsRef.current?.focus();
  };

  const handleClose = (): void => {
    toggleIsOpen();
    buttonRef.current?.focus();
  };

  const handleKeyboardNavigation = (event: KeyboardEvent): void => {
    event.preventDefault();

    if (!inputRefList.current) {
      return;
    }

    if (!isOpen) {
      handleOpen();
      return;
    } else if (event.key === KeyboardEventKey.Escape && isOpen) {
      handleClose();
      return;
    }

    const index = inputRefList.current.findIndex(x => x.current === event.target);
    let nextIndex = index;

    if (event.key === KeyboardEventKey.Home) {
      nextIndex = 0;
    } else if (event.key === KeyboardEventKey.End) {
      nextIndex = inputRefList.current.length - 1;
    } else if (event.key === KeyboardEventKey.ArrowDown && index < inputRefList.current.length - 1) {
      nextIndex = index + 1;
    } else if (event.key === KeyboardEventKey.ArrowUp && index > 0) {
      nextIndex = index - 1;
    } else if (event.key === KeyboardEventKey.Enter && index !== -1) {
      nextIndex = index;
    }
    if (nextIndex !== -1) {
      inputRefList.current[nextIndex].current?.focus();
      setCurrentIndex(nextIndex);
    }
  };

  useKeyboardEvent(dropdownRef, handleKeyboardNavigation, [
    KeyboardEventKey.ArrowDown,
    KeyboardEventKey.ArrowUp,
    KeyboardEventKey.End,
    KeyboardEventKey.Enter,
    KeyboardEventKey.Escape,
    KeyboardEventKey.Home,
  ]);

  useOutsideEvent(dropdownRef, () => isOpen && handleClose());

  const toggleClasses = classNames(
    styles.dropdown__toggle,
    !disabled && {
      [styles['dropdown__toggle--on-white']]: mode === DropdownMode.onwhite,
      [styles['dropdown__toggle--on-grey']]: mode === DropdownMode.ongrey,
      [styles['dropdown__toggle--on-blueberry']]: mode === DropdownMode.onblueberry,
      [styles['dropdown__toggle--on-cherry']]: mode === DropdownMode.oncherry,
      [styles['dropdown__toggle--transparent']]: transparent,
      [styles['dropdown__toggle--fluid']]: fluid,
      [styles['dropdown__toggle--open']]: isOpen,
    }
  );

  const contentClasses = classNames(styles.dropdown__content, isOpen && styles['dropdown__content--open']);

  const renderChildren = React.Children.map(children, (child, index) => (
    <li className={styles.dropdown__input} role="option" id={`${optionIdPrefix}-${index}`} aria-selected={index === currentIndex}>
      {React.isValidElement(child) && inputRefList.current && inputRefList.current[index]
        ? React.cloneElement(child as React.ReactElement, { ref: mergeRefs([child.props.ref, inputRefList.current[index]]) })
        : child}
    </li>
  ));

  return (
    <div className={styles.dropdown} ref={dropdownRef}>
      <span id={labelId} className={styles.dropdown__label}>
        {label}
      </span>
      <button
        type="button"
        onClick={(): false | void => !isOpen && handleOpen()}
        className={toggleClasses}
        ref={buttonRef}
        data-testid={testId}
        data-analyticsid={AnalyticsId.Dropdown}
        disabled={disabled}
        aria-labelledby={toggleLabelId}
        aria-haspopup="listbox"
        aria-expanded={isOpen}
      >
        <span id={toggleLabelId} className={styles.dropdown__toggle__label}>
          {placeholder}
        </span>
        <Icon
          color={disabled ? theme.palette.neutral700 : theme.palette.blueberry600}
          svgIcon={PlusSmall}
          className={styles.dropdown__icon}
          isHovered={!disabled && isHovered}
          size={IconSize.XSmall}
        />
      </button>
      <div className={contentClasses} style={{ width: fluid ? '100%' : `auto`, zIndex: zIndex }}>
        <ul
          className={styles.dropdown__options}
          role="listbox"
          aria-labelledby={labelId}
          tabIndex={-1}
          aria-activedescendant={typeof currentIndex !== 'undefined' ? `${optionIdPrefix}-${currentIndex}` : undefined}
          ref={optionsRef}
        >
          {renderChildren}
        </ul>
        {!noCloseButton && (
          <div className={styles.dropdown__close}>
            <Button onClick={handleClose} aria-expanded={isOpen}>
              {closeText}
            </Button>
          </div>
        )}
      </div>
    </div>
  );
};

export default Dropdown;
