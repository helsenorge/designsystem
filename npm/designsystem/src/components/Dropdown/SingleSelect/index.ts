import SingleSelectItem from './SingleSelectItem';
export * from './SingleSelectItem';
export default SingleSelectItem;
