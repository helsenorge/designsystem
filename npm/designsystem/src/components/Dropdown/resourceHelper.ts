import { LanguageLocales } from '../../constants';
import enGB from '../../resources/HN.Designsystem.Dropdown.en-GB.json';
import nbNO from '../../resources/HN.Designsystem.Dropdown.nb-NO.json';
import { HNDesignsystemDropdown } from '../../resources/Resources';

export const getResources = (language: LanguageLocales): HNDesignsystemDropdown => {
  switch (language) {
    case LanguageLocales.ENGLISH:
      return enGB;
    case LanguageLocales.NORWEGIAN:
    default:
      return nbNO;
  }
};
