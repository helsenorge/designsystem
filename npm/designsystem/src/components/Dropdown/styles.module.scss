@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';
@import '../../scss/screen-reader';

.dropdown {
  position: relative;

  &__label {
    @include sr-only;
  }

  &__toggle {
    display: flex;
    justify-content: space-between;
    align-items: center;
    cursor: pointer;
    border-radius: 0;
    border-width: 2px;
    border-style: solid;
    padding: 0 getSpacer(3xs) 0 getSpacer(s);
    font-family: inherit;
    font-size: $font-size-sm;
    line-height: 1.3rem; // Custom fordi teksten skal henge bedre sammen om den g√•r over to linjer

    @media (min-width: map.get($grid-breakpoints, md)) {
      padding: getSpacer(3xs) getSpacer(3xs) getSpacer(3xs) getSpacer(s);
      font-size: $font-size-md;
      line-height: 1.5rem;
    }

    &:focus {
      outline: none;
    }

    &:disabled {
      background-color: $neutral50;
      border-color: $neutral500;
    }

    &--on-white,
    &--on-grey {
      border-color: $neutral700;
      background-color: $white;

      &:hover {
        background-color: $neutral50;
        box-shadow: 0 0 0 1px $neutral700;
      }

      &:focus {
        border-color: $black;
        box-shadow: 0 0 0 1px $black;
      }
    }

    &--on-blueberry {
      border-color: $blueberry500;
      background-color: $white;

      &:hover {
        background-color: $blueberry50;
        box-shadow: 0 0 0 1px $blueberry500;
      }

      &:focus {
        border-color: $black;
        box-shadow: 0 0 0 1px $black;
      }
    }

    &--on-cherry {
      border-color: $cherry500;
      background-color: $white;

      &:hover {
        background-color: $cherry50;
        box-shadow: 0 0 0 1px $cherry500;
      }

      &:focus {
        border-color: $black;
        box-shadow: 0 0 0 1px $black;
      }
    }

    &--open#{&}--on-white,
    &--open#{&}--on-grey {
      background-color: $neutral50;

      &:hover {
        background-color: $white;
      }
    }

    &--open#{&}--on-blueberry {
      background-color: $blueberry50;

      &:hover {
        background-color: $white;
      }
    }

    &--open#{&}--on-cherry {
      background-color: $cherry50;

      &:hover {
        background-color: $white;
      }
    }

    &--transparent {
      background-color: transparent;
    }

    &--transparent#{&}--on-grey {
      &:hover {
        background-color: $neutral100;
      }
    }

    &--transparent#{&}--on-blueberry {
      &:hover {
        background-color: $blueberry100;
      }
    }

    &--transparent#{&}--on-cherry {
      &:hover {
        background-color: $cherry100;
      }
    }

    &--transparent#{&}--open#{&}--on-grey {
      background-color: $neutral100;

      &:hover {
        background-color: transparent;
      }
    }

    &--transparent#{&}--open#{&}--on-blueberry {
      background-color: $blueberry100;

      &:hover {
        background-color: transparent;
      }
    }

    &--transparent#{&}--open#{&}--on-cherry {
      background-color: $cherry100;

      &:hover {
        background-color: transparent;
      }
    }

    &--fluid {
      width: 100%;
    }

    &__label {
      padding: getSpacer(3xs) 0;

      @media (min-width: map.get($grid-breakpoints, md)) {
        padding: getSpacer(2xs) 0;
      }
    }
  }

  &__icon {
    margin-left: getSpacer(2xs);
  }

  &__content {
    display: none;
    position: absolute;
    background-color: $white;
    margin-top: getSpacer(2xs);
    box-shadow: 0 0 0 2px $blueberry500;
    z-index: 1;

    &--open {
      display: block;
    }
  }

  &__options {
    overflow-x: hidden;
    overflow-y: auto;
    max-height: 384px;
    list-style: none;
    padding: 0;
    margin: 0;

    @media (min-width: map.get($grid-breakpoints, md)) {
      max-height: 464px;
    }
  }

  &__close {
    padding: getSpacer(s);
  }

  &__input {
    border-bottom: 1px solid $neutral300;
    padding: getSpacer(2xs) getSpacer(s) getSpacer(2xs) getSpacer(s);

    @media (min-width: map.get($grid-breakpoints, md)) {
      padding: getSpacer(s);
    }

    &:last-child {
      border-bottom: none;
    }
  }
}
