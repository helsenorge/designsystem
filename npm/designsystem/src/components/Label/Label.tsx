import React, { FunctionComponent } from 'react';

import cn from 'classnames';

import { Sublabel, SublabelProps } from './SubLabel';
import { AnalyticsId, FormOnColor } from '../../constants';
import { isComponent } from '../../utils/component';
import FormFieldTag, { FormFieldTagLevel } from '../FormFieldTag';
import Spacer from '../Spacer';
import StatusDot, { StatusDotProps } from '../StatusDot';

import styles from './styles.module.scss';

export type LabelText = {
  hideFromScreenReader?: boolean;
  text: string;
  type?: 'normal' | 'subdued';
};

export type LabelTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'label' | 'p';

export interface LabelProps {
  /** Component shown after label - discourage use of this */
  afterLabelChildren?: React.ReactNode;
  /** Sets the content of the Label */
  children?: React.ReactNode;
  /** Sets a tag that describes whether the form field is required or optional */
  formFieldLevel?: FormFieldTagLevel;
  /** Adds custom classes to the label tag. */
  labelClassName?: string;
  /** Adds custom classes to the label text. */
  labelTextClassName?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Id that is put on the "for" attribute of the label */
  htmlFor?: string;
  /** Changes the underlying element of the label */
  htmlMarkup?: LabelTags;
  /** Id som plasseres p√• <label/> */
  labelId?: string;
  /** Array of main label strings. Can be of type semibold or normal */
  labelTexts?: LabelText[];
  /** Array of sublabel strings. Can be of type semibold or normal */
  onColor?: keyof typeof FormOnColor;
  /** StatusDot placed underneath the last sublabel */
  statusDot?: React.ReactNode;
  /** Sublabel component */
  sublabel?: React.ReactNode;
  /** Adds custom classes to the div wrapping the sublabels. */
  sublabelWrapperClassName?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const getLabelText = (label: React.ReactNode): string => {
  let allLabelText = '';

  if (isComponent<LabelProps>(label, Label)) {
    label.props.labelTexts?.forEach(labelText => {
      allLabelText += !labelText.hideFromScreenReader ? labelText.text : '';
    });
  }

  return allLabelText;
};

export const renderLabel = (label: React.ReactNode, inputId: string, onColor: FormOnColor, markup?: LabelTags): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            htmlMarkup: markup || 'label',
            onColor,
          })
        : typeof label === 'string' && <Label labelTexts={[{ text: label, type: 'normal' }]} htmlFor={inputId} onColor={onColor} />}
    </>
  );
};

export const renderLabelAsParent = (
  label: React.ReactNode,
  children: React.ReactNode,
  inputId: string,
  onColor: FormOnColor,
  labelClassName?: string,
  labelTextClassName?: string,
  sublabelWrapperClassName?: string,
  large?: boolean,
  markup?: LabelTags
): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            onColor,
            children: children,
            labelClassName: cn(labelClassName, label.props.labelClassName),
            labelTextClassName: labelTextClassName,
            htmlMarkup: markup || 'label',
            sublabelWrapperClassName: sublabelWrapperClassName,
            sublabel: large ? undefined : label.props.sublabel,
            statusDot: large ? undefined : label.props.statusDot,
          })
        : typeof label === 'string' && (
            <Label
              labelTexts={[{ text: label, type: 'subdued' }]}
              htmlFor={inputId}
              onColor={onColor}
              htmlMarkup={markup || 'label'}
              labelClassName={labelClassName}
              labelTextClassName={labelTextClassName}
              sublabelWrapperClassName={sublabelWrapperClassName}
            >
              {children}
            </Label>
          )}
    </>
  );
};

const Label: FunctionComponent<LabelProps> = ({
  afterLabelChildren,
  children,
  className,
  formFieldLevel,
  htmlFor,
  htmlMarkup = 'label',
  labelClassName,
  labelTextClassName,
  labelId,
  labelTexts,
  onColor = FormOnColor.onwhite,
  statusDot,
  sublabel,
  sublabelWrapperClassName,
  testId,
}) => {
  const hasChildren = children && typeof children !== 'undefined';
  const labelWrapperClasses = cn(
    styles['label-wrapper'],
    { [styles['label-wrapper--no-bottom-margin']]: hasChildren, [styles['label-wrapper--after-label-children']]: afterLabelChildren },
    className
  );

  const mapLabels = (): React.ReactNode => {
    if (typeof labelTexts === 'undefined') return null;

    return labelTexts.map((labelText, index) => {
      const labelClasses = cn(
        styles.label,
        {
          [styles['label--subdued']]: labelText.type === 'subdued',
          [styles['label--on-dark']]: onColor === FormOnColor.ondark,
        },
        labelTextClassName
      );
      return (
        <span aria-hidden={labelText.hideFromScreenReader} className={labelClasses} key={index}>
          {labelText.text}
        </span>
      );
    });
  };
  const CustomTag = htmlMarkup;

  return (
    <div className={labelWrapperClasses}>
      <div>
        <CustomTag className={labelClassName} id={labelId} htmlFor={htmlFor} data-testid={testId} data-analyticsid={AnalyticsId.Label}>
          <span className={styles['label-content-wrapper']}>
            {children}
            <span className={styles.label__texts}>{mapLabels()}</span>
          </span>
        </CustomTag>
        <div className={sublabelWrapperClassName}>
          {formFieldLevel && <FormFieldTag level={formFieldLevel} />}
          {sublabel &&
            isComponent<SublabelProps>(sublabel, Sublabel) &&
            React.cloneElement(sublabel, {
              onColor: onColor as FormOnColor,
            })}
          {statusDot && isComponent<StatusDotProps>(statusDot, StatusDot) && (
            <>
              <Spacer size={'3xs'} />
              {React.cloneElement(statusDot, {
                onColor: onColor === FormOnColor.ondark ? 'ondark' : 'onwhite',
              })}
            </>
          )}
        </div>
      </div>
      {afterLabelChildren && <div className={styles['after-label-children']}>{afterLabelChildren}</div>}
    </div>
  );
};

export default Label;
