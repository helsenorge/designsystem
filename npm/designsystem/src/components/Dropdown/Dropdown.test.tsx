import React from 'react';

import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Label from '../Label';
import RadioButton from '../RadioButton';

import Dropdown from './';

import '../../__mocks__/uuid';

describe('Gitt at Dropdown skal vises vanlig', (): void => {
  describe('Når Dropdownen har vanlig innhold', (): void => {
    test('Så kan man klikke og se innholdet', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });

      await userEvent.click(button);

      const options = screen.getByLabelText('Ta et valg');
      expect(options).toHaveFocus();

      const contents = within(options).getByRole('heading', { name: 'Innhold i Dropdown' });
      expect(contents).toBeVisible();
    });

    test('Så ser Dropdown ut slik den skal', (): void => {
      const { container } = render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Når Dropdownen har lukkeknapp', (): void => {
    test('Så kan man lukke dropdownen med knappen', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp" closeText="Lukk">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'false');
      await userEvent.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'true');
      const close = screen.getByRole('button', { name: 'Lukk' });
      await userEvent.click(close);
      expect(close).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når Dropdownen har open prop', (): void => {
    test('Så er den åpen', (): void => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp" open>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'true');
    });
  });

  describe('Når Dropdownen har disabled prop', (): void => {
    test('Så er den disabled', (): void => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp" disabled>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toBeDisabled();
    });
  });

  describe('Når Dropdownen har noCloseButton prop', (): void => {
    test('Så er det ingen lukkeknapp', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp" closeText="Lukk" noCloseButton>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      await userEvent.click(button);

      const close = screen.queryByRole('button', { name: 'Lukk' });
      expect(close).not.toBeInTheDocument();
    });
  });

  describe('Når man klikker utenfor dropdownen mens den er åpen', (): void => {
    test('Så lukkes Dropdownen', async (): Promise<void> => {
      render(
        <>
          <button>Knapp utenfor dropdown</button>
          <Dropdown label="Ta et valg" placeholder="Knapp" open>
            <h2>Innhold i Dropdown</h2>
          </Dropdown>
        </>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'true');

      const outsideButton = screen.getByRole('button', { name: 'Knapp utenfor dropdown' });
      await userEvent.click(outsideButton);

      expect(button).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når man klikker på Escape mens Dropdown er åpen', (): void => {
    test('Så lukkes Dropdownen', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      await userEvent.click(button);
      await userEvent.keyboard('{Escape}');

      expect(button).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når man klikker på Enter mens Dropdown er lukket', (): void => {
    test('Så åpnes Dropdownen', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      // Åpne og lukke dropdownen slik at fokus er på knappen før vi klikker på tastaturet
      await userEvent.click(button);
      await userEvent.click(button);
      await userEvent.keyboard('{enter}');

      expect(button).toHaveAttribute('aria-expanded', 'true');
    });
  });

  describe('Når man klikker på Home mens Dropdown er åpen', (): void => {
    test('Så flyttes fokus til første radioknapp, og man kan bruke tastaturet for å gå nedover, men ikke oppover', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <RadioButton label={<Label labelTexts={[{ text: 'Førstevalg' }]} />} inputId="radio-1" value="radio-1" />
          <RadioButton label={<Label labelTexts={[{ text: 'Andrevalg' }]} />} inputId="radio-2" value="radio-2" />
          <RadioButton label={<Label labelTexts={[{ text: 'Tredjevalg' }]} />} inputId="radio-3" value="radio-3" />
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      await userEvent.click(button);
      await userEvent.keyboard('{home}');

      const radio1 = screen.getByLabelText('Førstevalg');
      expect(radio1).toHaveFocus();

      await userEvent.keyboard('{arrowup}');
      expect(radio1).toHaveFocus();

      await userEvent.keyboard('{arrowdown}');
      const radio2 = screen.getByLabelText('Andrevalg');
      expect(radio2).toHaveFocus();
    });
  });

  describe('Når man klikker på End mens Dropdown er åpen', (): void => {
    test('Så flyttes fokus til sist radioknapp, og man kan bruke tastaturet for å gå oppover, men ikke nedover', async (): Promise<void> => {
      render(
        <Dropdown label="Ta et valg" placeholder="Knapp">
          <RadioButton label={<Label labelTexts={[{ text: 'Førstevalg' }]} />} inputId="radio-1" value="radio-1" />
          <RadioButton label={<Label labelTexts={[{ text: 'Andrevalg' }]} />} inputId="radio-2" value="radio-2" />
          <RadioButton label={<Label labelTexts={[{ text: 'Tredjevalg' }]} />} inputId="radio-3" value="radio-3" />
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      await userEvent.click(button);
      await userEvent.keyboard('{end}');

      const radio3 = screen.getByLabelText('Tredjevalg');
      expect(radio3).toHaveFocus();

      await userEvent.keyboard('{arrowdown}');
      expect(radio3).toHaveFocus();

      await userEvent.keyboard('{arrowup}');

      const radio2 = screen.getByLabelText('Andrevalg');
      expect(radio2).toHaveFocus();
    });
  });
});
