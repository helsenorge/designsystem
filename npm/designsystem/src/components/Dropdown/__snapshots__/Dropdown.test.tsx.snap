// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at Dropdown skal vises vanlig > Når Dropdownen har vanlig innhold > Så ser Dropdown ut slik den skal 1`] = `
<div>
  <div
    class="dropdown"
  >
    <span
      class="dropdown__label"
      id="testid-9"
    >
      Ta et valg
    </span>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="testid-10"
      class="dropdown__toggle dropdown__toggle--on-white"
      data-analyticsid="dropdown"
      type="button"
    >
      <span
        class="dropdown__toggle__label"
        id="testid-10"
      >
        Knapp
      </span>
      <svg
        aria-hidden="true"
        class="hnds-style-icon dropdown__icon"
        data-analyticsid="icon"
        fill="#126F87"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M24 14.18v8.556h8.558v2.527H24v8.558h-2.526l-.001-8.558h-8.557v-2.526h8.557v-8.558H24z"
        />
      </svg>
    </button>
    <div
      class="dropdown__content"
      style="width: auto; z-index: 11000;"
    >
      <ul
        aria-labelledby="testid-9"
        class="dropdown__options"
        role="listbox"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="dropdown__input"
          id="testid-11-0"
          role="option"
        >
          <h2>
            Innhold i Dropdown
          </h2>
        </li>
      </ul>
      <div
        class="dropdown__close"
      >
        <button
          aria-expanded="false"
          class="button-wrapper"
          data-analyticsid="button"
          type="button"
        >
          <span
            class="button button--normal"
          >
            <span
              class="button__text"
            >
              <span>
                Lukk
              </span>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
